### PR TITLE
Implement (de)materialize

### PR DESCRIPTION
**Pull request by maikel**

### CODE REVIEW

File: include/exec/materialize.hpp [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+        using __materialize_error = completion_signatures<set_value_t(set_error_t, _E)>;
+
+        template <class _Env>
+        using __completion_signatures_for_t = make_completion_signatures<
+          _Sender,
+          _Env,
+          completion_signatures<>,
+          __materialize_value,
+          __materialize_error,
+          completion_signatures<set_value_t(set_stopped_t)>>;
```
Comment (ericniebler) [inline include/exec/materialize.hpp:92]: Add this one only if the input sender can complete with stopped.

File: include/exec/materialize.hpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+        _Receiver __upstream_;
+
+        template <__completion_tag _Tag, __decays_to<__t> _Self, class... _Args>
+          requires tag_invocable<set_value_t, _Receiver&&, _Tag, _Args...>
+        friend void tag_invoke(_Tag tag, _Self&& __self, _Args&&... __args) noexcept {
+          set_value((_Receiver&&) __self.__upstream_, tag, (_Args&&) __args...);
+        }
+
+        template <__decays_to<__t> _Self>
+        friend env_of_t<_Receiver> tag_invoke(get_env_t, _Self&& __self) noexcept {
```
Comment (ericniebler) [inline include/exec/materialize.hpp:45]:
```suggestion
        template <same_as<__t> _Self>
        friend env_of_t<_Receiver> tag_invoke(get_env_t, const _Self& __self) noexcept {
```

File: include/exec/materialize.hpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        }
+
+        template <__one_of<set_stopped_t, set_error_t> _Tag, __decays_to<__t> _Self, class... _Args>
+          requires tag_invocable<_Tag, _Receiver&&, _Args...>
+        friend void tag_invoke(_Tag tag, _Self&& __self, _Args&&... __args) noexcept {
+          tag((_Receiver&&) __self.__upstream_, (_Args&&) __args...);
+        }
+
+        template <__decays_to<__t> _Self>
+        friend env_of_t<_Receiver> tag_invoke(get_env_t, _Self&& __self) noexcept {
```
Comment (ericniebler) [inline include/exec/materialize.hpp:144]:
```suggestion
        template <same_as<__t> _Self>
        friend env_of_t<_Receiver> tag_invoke(get_env_t, const _Self& __self) noexcept {
```

File: include/exec/materialize.hpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        friend connect_result_t<_Sender, __receiver_t<_Receiver>>
+          tag_invoke(connect_t, _Self&& __self, _Receiver&& __receiver) noexcept(
+            __nothrow_connectable<__copy_cvref_t<_Self, _Sender>, __receiver_t<_Receiver>>) {
+          return connect(
+            ((_Self&&) __self).__sender_, __receiver_t<_Receiver>{(_Receiver&&) __receiver});
+        }
+
+        template <class _Tag, class... _Args>
+          requires __completion_tag<decay_t<_Tag>>
+        using __dematerialize_value = completion_signatures<decay_t<_Tag>(_Args...)>;
```
Comment (ericniebler) [inline include/exec/materialize.hpp:180]: Very nice.

**Review by ericniebler — COMMENTED** [review]
Terrific. You're on a roll. :-) Just one small change.

File: include/exec/materialize.hpp (hunk 1 above)
Comment (maikel) [reply include/exec/materialize.hpp:92]: But isnt that what automatically happens? I am testing this

```cpp
  auto just_ = materialize(just());
  static_assert(sender<decltype(just_)>);
  static_assert(__v<__equivalent<
                  completion_signatures_of_t<decltype(just_)>,
                  completion_signatures<set_value_t(set_value_t)>>>);
```

I mean, I can add 

```cpp
          __if_c<
            sends_stopped<_Sender, _Env>,
            completion_signatures<set_value_t(set_stopped_t)>,
            completion_signatures<>>>
```

But I don't understand the rationale

Comment (ericniebler) [reply include/exec/materialize.hpp:92]: Oh yeah I forgot

**Comment by kirkshoop** [comment]
Wow this is beautiful 🤩

**Pull request merged**